compiler/rustc_hir_typeck/src/callee.rs

```diff
@@ -722,12 +731,57 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err.emit()
     }
 
+    /// Enforces that things being called actually are callable
+    #[instrument(skip(self, arguments))]
+    pub(super) fn check_callable(
+        &self,
+        hir_id: hir::HirId,
+        span: Span,
+        value: Ty<'tcx>,
+        arguments: impl IntoIterator<Item = Ty<'tcx>>,
+    ) {
+        if value.references_error() {
+            return;
+        }
+
+        let context = self.tcx.hir().enclosing_body_owner(hir_id);
+        let kind = self.tcx.def_kind(context.to_def_id());
+        debug_assert_ne!(kind, DefKind::ConstParam);
+
+        let ty::FnDef(did, substs) = *value.kind() else {
+            trace!("not a function item");
+            return;
+        };
+        trace!(?substs);
+
+        // Tuple struct/variant constructors are not really function calls, they are always fine.
+        if let DefKind::Ctor(_, def::CtorKind::Fn) = self.tcx.def_kind(did) {
```

Comment:
Is this (and other early-returns below) for perf? Why not just register the predicate anyways even if we know it'll trivially hold? The solver(s) (ideally) should be able to handle a few extra trait preds.